HW_Type_Protectors.ts

```diff
@@ -0,0 +1,99 @@
+type EmployeeStatus = 'active' | 'inactive' | 'unpaid_leave';
+
+interface PreHiredEmployee {
+    firstName: string;
+    lastName: string;
+    salary: number;
+    bankAccount: string;
+}
+
+interface Employee extends PreHiredEmployee {
+    paymentInfo: string;
+    status: EmployeeStatus;
+    department: Department;
+}
+
+interface Budget {
+    debit: number;
+    credit: number;
+}
+
+interface Department {
+    name: string;
+    domain: string;
+    employees: Employee[];
+    budget: Budget;
+
+    calculateBalance(): number;
+    addEmployee(employee: Employee): void;
+    removeEmployee(employee: Employee): void;
+    promoteToEmployee(preHired: PreHiredEmployee): Employee;
+}
+
+class Company {
+    name: string;
+    departments: Department[];
+    preHiredEmployees: PreHiredEmployee[];
+    allEmployees: (Employee | PreHiredEmployee)[];
+
+    constructor(name: string, departments: Department[], preHiredEmployees: PreHiredEmployee[]) {
+        this.name = name;
+        this.departments = departments;
+        this.preHiredEmployees = preHiredEmployees;
+        this.allEmployees = [].concat(...departments.map(d => d.employees), preHiredEmployees);
+    }
+}
+
+class Accounting implements Department {
+    name: string;
+    domain: string;
+    employees: Employee[];
+    budget: Budget;
+    balance: number;
+
+    constructor(name: string, domain: string, budget: Budget) {
+        this.name = name;
+        this.domain = domain;
+        this.employees = [];
+        this.budget = budget;
+        this.balance = this.calculateBalance();
+    }
+
+    calculateBalance(): number {
+        return this.budget.credit - this.budget.debit;
+    }
+
+    addEmployee(employee: Employee): void {
+        this.employees.push(employee);
+        this.updateBalanceForEmployee(employee);
+    }
+
+    removeEmployee(employee: Employee): void {
+        this.employees = this.employees.filter(emp => emp !== employee);
+    }
+
+    promoteToEmployee(preHired: PreHiredEmployee): Employee {
```

Comment:
In the method 'addEmployee', we need to handle the addition of prehired employees to the department. You instead of this simplifies the task and avoids the use of type guards.